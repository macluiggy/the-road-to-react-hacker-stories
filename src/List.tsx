import styled from 'styled-components';
import { ReactComponent as Check } from './check.svg';

import React from 'react';

const List = 
        ({list, onRemoveItem}: listProps) => (
                    console.log('B:list') ||
                <>
                    {list.map(item => (
                        <Item 
                            key={item.objectID} 
                            item={item}
                            onRemoveItem={onRemoveItem} 
                        />
                    ))}
                </>
        )

const StyledItem = styled.div`
display: flex;
align-items: center;
padding-bottom: 5px;
`;

const StyledColumn = styled.span`
padding: 0 5px;
white-space: nowrap;
overflow: hidden;
white-space: nowrap;
text-overflow: ellipsis;
a {
color: red;
}
width: ${props => props.width};
`;




const Item = ({ item, onRemoveItem }: listProps) => (
  <StyledItem>
    <span style={{ width: '40%' }}>
      <a href={item.url}>{item.title}</a>
    </span>
    <StyledColumn width='30%'>{item.author}</StyledColumn>
    <StyledColumn width="10%">{item.num_comments}</StyledColumn>
    <StyledColumn width="10%">{item.points}</StyledColumn>
    <StyledColumn width="10%">
      <button
       type="button"
       onClick={() => onRemoveItem(item)}
       className="button button_small"
      >
        <Check height="18px" width="18px" />
      </button>
    </StyledColumn>
  </StyledItem>
);

export default List;