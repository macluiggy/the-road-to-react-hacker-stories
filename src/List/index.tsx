import styled from 'styled-components';
import { ReactComponent as Check } from '../check.svg';
import { sortBy } from 'lodash';

import React from 'react';

const SORTS = {
    NONE: list => list,
    TITLE: list => sortBy(list, 'title'),
    AUTHOR: list => sortBy(list, 'author'),
    COMMENT: list => sortBy(list, 'num_comments').reverse(),
    POINT: list => sortBy(list, 'points').reverse(),
}
const List = 
        ({list, onRemoveItem}: listProps) => {
            const [sort, setSort] = React.useState({
                sortKey: 'NONE',
                isReverse: false,
            });

            const handleSort = sortKey => {
                const isReverse = sort.sortKey === sortKey && !sort.isReverse;
                setSort({ sortKey, isReverse });
            }

            const sortFunction = SORTS[sort.sortKey];
            const sortedList = sort.isReverse
                ? sortFunction(list).reverse()
                : sortFunction(list);

            return (console.log('B:list')) || (
                <div>
                    <div style={{ display: 'flex' }}>
                        <span 
                        style={{ width: '40%' }}
                        type='button' onClick={() => handleSort('TITLE')} >
                            <button style={{
                                backgroundColor: sort.sortKey === 'TITLE' ? 'red' : 'white'
                            }} >
                                Title
                            </button>
                        </span>
                        <span 
                        style={{ width: '30%' }}
                        type='button' onClick={() => handleSort('AUTHOR')} >
                            <button style={{
                                backgroundColor: sort.sortKey === 'AUTHOR' ? 'red' : 'white',
                            }} >
                                Author
                            </button>
                        </span>
                        <span 
                        style={{ width: '10%' }}
                        type='button' onClick={() => handleSort('COMMENT')} >
                            <button style={{
                                backgroundColor: sort.sortKey === 'COMMENT' ? 'red' : 'white',
                            }} >
                                Comments
                            </button>
                        </span>
                        <span 
                        style={{ width: '10%' }}
                        type='button' onClick={() => handleSort('POINT')} >
                            <button style={{
                                backgroundColor: sort.sortKey === 'POINT' ? 'red' : 'white',
                            }} >
                                Points
                            </button>
                        </span>
                        <span 
                        style={{ width: '10%' }}>Actions</span>
                    </div>

                    {sortedList.map(item => (
                        <Item 
                            key={item.objectID} 
                            item={item}
                            onRemoveItem={onRemoveItem} 
                        />
                    ))}
                </div>
            )
        }

const StyledItem = styled.div`
display: flex;
align-items: center;
padding-bottom: 5px;
`;

const StyledColumn = styled.span`
padding: 0 5px;
white-space: nowrap;
overflow: hidden;
white-space: nowrap;
text-overflow: ellipsis;
a {
color: red;
}
width: ${props => props.width};
`;




export const Item = ({ item, onRemoveItem }: listProps) => (
  <StyledItem>
    <span style={{ width: '40%' }}>
      <a href={item.url}>{item.title}</a>
    </span>
    <StyledColumn width='30%'>{item.author}</StyledColumn>
    <StyledColumn width="10%">{item.num_comments}</StyledColumn>
    <StyledColumn width="10%">{item.points}</StyledColumn>
    <StyledColumn width="10%">
      <button
       type="button"
       onClick={() => onRemoveItem(item)}
       className="button button_small"
      >
        <Check height="18px" width="18px" />
      </button>
    </StyledColumn>
  </StyledItem>
);

export default List;