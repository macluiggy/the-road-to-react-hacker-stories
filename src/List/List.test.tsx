import React from 'react';
import renderer from 'react-test-renderer';

import List, { Item } from './index.tsx';

describe('List', () => {
  const list = [
    {
      title: 'React',
      url: 'https://reactjs.org/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: 'Redux',
      url: 'https://redux.js.org/',
      author: 'Dan Abramov, Andrew Clark',
      num_comments: 2,
      points: 5,
      objectID: 1,
    },
  ]


  //crea la funcion jest para que tome el lugar de la funcion del evento
  //cuando sea llamada
  const handleRemoveItem = jest.fn()

  //crea un component jest para el componente List
  const component = renderer.create(
    <List list={list} onRemoveItem={handleRemoveItem} />
  );

  it('renders two items', () => {

    //Dado que hay mas de un item dentro del cual cada item tiene el atributo 'a'
    //se debe usar findAllByType
    //se espera que el primer elemento a tenga su url igual a list[0].url
    //y el segunfo sea igaul a list[1].url
    expect(component.root.findAllByType('a')[0].props.href).toEqual('https://reactjs.org/')
    expect(component.root.findAllByType('a')[1].props.href).toEqual('https://redux.js.org/')
    //encuentra de ese componente el elemento react que sea boton y llama 2 veces
    //a la funcion que se encuentra en el componente List
    component.root.findAllByType('button')[0].props.onClick();
    component.root.findAllByType('button')[1].props.onClick();

    //haz el test que espera que la funcion haya sido llamada 2 vez
    expect(handleRemoveItem).toHaveBeenCalledTimes(2);

    //se espera que la cantidad de Item que se mapean en el componente
    //List sean los mismos que la variable list
    expect(component.root.findAllByType(Item).length).toEqual(2)
  })

  test('renders snapshot for List component', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot()
  })
})

describe('Item', () => {
  const item = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  };
  const handleRemoveItem = jest.fn();

  let component;

  beforeEach(() => {
    component = renderer.create(
      <Item item={item} onRemoveItem={handleRemoveItem} />
    )
  })

  it('render all properties', () => {
    const component = renderer.create(<Item item={item} />);
    expect(component.root.findByType('a').props.href).toEqual(
      'https://reactjs.org/'
    )
    expect(
      component.root.findAllByType('span')[1].props.children
    ).toEqual('Jordan Walke')
    expect(
      component.root.findAllByProps({children: 'Jordan Walke'}).length
    ).toEqual(2)
    expect(
      component.root.findAllByType('span')[2].props.children
    ).toEqual(3)
    //Se espera encontrar un solo tipo de elemento 'a' y que su children
    //sea igual a 'React'
    expect(
      component.root.findByType('a').props.children
    ).toEqual('React');
    //este test es basicamente igual al anterior, se espera el mismo resultado
    // solo que aqui se usa 'findAllByProps' y un objeto que contiene la prop deseada
    expect(
      component.root.findAllByProps({ children: 'React' })
        .length
    ).toEqual(1)
  });

  it('calls onRemoveItem on button click', () => {
    const handleRemoveItem = jest.fn();//crea un simulacro de la funcion, esta funcion
    //puede tener cualquier nombre, ya que la funcion que emula es an√≥nima

    const component = renderer.create(
      <Item  item={item} onRemoveItem={handleRemoveItem} />//crea un componente jest
    )

    //encuentra de ese componente el elemento react que sea boton y llama a la funcion
    //que se encuentra en este elemento
    component.root.findByType('button').props.onClick();

    //haz el test que espera que la funcion haya sido llamada 1 vez
    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
    //se hace el test de que la funcion haya sido invocada con el argumento item
    expect(handleRemoveItem).toHaveBeenCalledWith(item);

    //testea que la cantidad de componentes Item sea igual a uno
    expect(component.root.findAllByType(Item).length).toEqual(1)
  })
  it(`poinst equal to ${item.points}`, () => {
    const component = renderer.create(<Item item={item} />);
      expect(component.root.findAllByType('span')[3].props.children)
      .toEqual(4)
    });

  //este test crea un snapshot de la estructura HTML del componente la primera vez que se hace
  //un test, de esta manera si se llega a cambiar la estructura, y luego se corre otro test,
  //este fallara y mostrara los cambios que hay con respecto a la ultima snapshot
  test('renders snapshot', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot()
  })
})