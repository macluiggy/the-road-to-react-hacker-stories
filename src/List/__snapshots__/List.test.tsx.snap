// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
<div
  className="sc-bdnxRM fBLZFc"
>
  <span
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    className="sc-gtsrHT iBufAq"
    width="30%"
  >
    Jordan Walke
  </span>
  <span
    className="sc-gtsrHT iBtGzw"
    width="10%"
  >
    3
  </span>
  <span
    className="sc-gtsrHT iBtGzw"
    width="10%"
  >
    4
  </span>
  <span
    className="sc-gtsrHT iBtGzw"
    width="10%"
  >
    <button
      className="button button_small"
      onClick={[Function]}
      type="button"
    >
      <svg
        height="18px"
        width="18px"
      >
        check.svg
      </svg>
    </button>
  </span>
</div>
`;

exports[`List renders snapshot for List component 1`] = `
<div>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <span
      onClick={[Function]}
      style={
        Object {
          "width": "40%",
        }
      }
      type="button"
    >
      <button
        style={
          Object {
            "backgroundColor": "white",
          }
        }
      >
        Title
      </button>
    </span>
    <span
      onClick={[Function]}
      style={
        Object {
          "width": "30%",
        }
      }
      type="button"
    >
      <button
        style={
          Object {
            "backgroundColor": "white",
          }
        }
      >
        Author
      </button>
    </span>
    <span
      onClick={[Function]}
      style={
        Object {
          "width": "10%",
        }
      }
      type="button"
    >
      <button
        style={
          Object {
            "backgroundColor": "white",
          }
        }
      >
        Comments
      </button>
    </span>
    <span
      onClick={[Function]}
      style={
        Object {
          "width": "10%",
        }
      }
      type="button"
    >
      <button
        style={
          Object {
            "backgroundColor": "white",
          }
        }
      >
        Points
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Actions
    </span>
  </div>
  <div
    className="sc-bdnxRM fBLZFc"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://reactjs.org/"
      >
        React
      </a>
    </span>
    <span
      className="sc-gtsrHT iBufAq"
      width="30%"
    >
      Jordan Walke
    </span>
    <span
      className="sc-gtsrHT iBtGzw"
      width="10%"
    >
      3
    </span>
    <span
      className="sc-gtsrHT iBtGzw"
      width="10%"
    >
      4
    </span>
    <span
      className="sc-gtsrHT iBtGzw"
      width="10%"
    >
      <button
        className="button button_small"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>
  <div
    className="sc-bdnxRM fBLZFc"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://redux.js.org/"
      >
        Redux
      </a>
    </span>
    <span
      className="sc-gtsrHT iBufAq"
      width="30%"
    >
      Dan Abramov, Andrew Clark
    </span>
    <span
      className="sc-gtsrHT iBtGzw"
      width="10%"
    >
      2
    </span>
    <span
      className="sc-gtsrHT iBtGzw"
      width="10%"
    >
      5
    </span>
    <span
      className="sc-gtsrHT iBtGzw"
      width="10%"
    >
      <button
        className="button button_small"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>
</div>
`;
